scripts/migration-library.sh

```diff
@@ -0,0 +1,575 @@
+    if [ ! -f "${CK8S_CONFIG_PATH}/${FILE}" ]; then
+      log_error "error: \"${FILE}\" is not a file"
+      pass="false"
+    fi
+  done
+
+  if ! check_sops "${CK8S_CONFIG_PATH}/secrets.yaml"; then
+      log_error "error: \"secrets.yaml\" is not encrypted"
+      pass="false"
+  fi
+
+  if [[ "${pass}" = "false" ]]; then
+    exit 1
+  fi
+
+  for prefix in sc wc; do
+    if check_sops "${CK8S_CONFIG_PATH}/.state/kube_config_${prefix}.yaml"; then
+      CONFIG["${prefix}-kubeconfig"]="encrypted"
+    else
+      CONFIG["${prefix}-kubeconfig"]="unencrypted"
+    fi
+
+    log_info "using ${prefix} kubeconfig ${CONFIG["${prefix}-kubeconfig"]}"
+  done
+}
+
+# usage: check_version <sc|wc> <prepare|apply>
+check_version() {
+  case "${1:-}" in
+  sc|wc)
+    ;;
+  *)
+    log_fatal "usage: check_version <sc|wc> <prepare|apply>"
+    ;;
+  esac
+
+  if [ "${VERSION["${1}-config"]}" = "any" ]; then
+    return
+  fi
+
+  local version="${HERE##"${ROOT}/migration/v"}"
+  local major="${version%%.*}"
+  local minor="${version##*.}"
+
+  case "${2:-}" in
+  prepare)
+    if [ $((major - 1)) -eq "${VERSION["${1}-config-major"]}" ]; then
+      if [ $((minor)) -eq 0 ]; then
+        log_info "valid upgrade path to next major version v${major}.${minor}"
+      else
+        log_fatal "invalid upgrade path to major version v${major}.${minor}"
+      fi
+    elif [ $((major)) -eq "${VERSION["${1}-config-major"]}" ]; then
+      if [ $((minor - 1)) -eq "${VERSION["${1}-config-minor"]}" ]; then
+        log_info "valid upgrade path to next minor version v${major}.${minor}"
+      elif [ $((minor)) -eq "${VERSION["${1}-config-minor"]}" ]; then
+        log_info "valid upgrade path to patch version v${major}.${minor}"
+      else
+        log_fatal "invalid upgrade path to minor version v${major}.${minor}"
+      fi
+    else
+      log_fatal "invalid upgrade path to version v${major}.${minor}"
+    fi
+    ;;
+  apply)
+    if [ $((major)) -eq "${VERSION["${1}-config-major"]}" ] && [ $((minor)) -eq "${VERSION["${1}-config-minor"]}" ]; then
+      log_info "valid upgrade path to version v${major}.${minor}"
+    else
+      log_fatal "invalid upgrade path to version v${major}.${minor}"
+    fi
+    ;;
+  *)
+    log_fatal "usage: check_version <sc|wc> <prepare|apply>"
+    ;;
+  esac
+}
+
+# --- kubectl functions ---
+
+# usage: kubectl_do <sc|wc> <args...>
+kubectl_do() {
+  case "${1:-}" in
+  sc|wc)
+    if [ "${CONFIG["${1}-kubeconfig"]}" = "encrypted" ]; then
+      kubectl --kubeconfig <(sops -d "${CK8S_CONFIG_PATH}/.state/kube_config_${1}.yaml") "${@:2}"
+    else
+      kubectl --kubeconfig "${CK8S_CONFIG_PATH}/.state/kube_config_${1}.yaml" "${@:2}"
+    fi
+    ;;
+  *)
+    log_fatal "usage: kubectl_* <sc|wc> <args...>"
+    ;;
+  esac
+}
+
+# usage: kubectl_delete <sc|wc> <resource> <namespace> <name>
+kubectl_delete() {
+  kubectl_do "${1:-}" delete "${2:-}" --namespace "${3:-}" "${4:-}" --ignore-not-found true
+}
+
+# --- helm functions ---
```

Comment:
As you have done a good job separating the fucntions, i would say spit them up into separate files. So that this file does not become so long :smile: